drivers/usb/device/usb_dc_rpi_pico.c

```diff
@@ -635,15 +635,27 @@ int usb_dc_ep_read_continue(uint8_t ep)
 		return -EINVAL;
 	}
 
-	bytes_received = state.setup_available ?
-			 sizeof(struct usb_setup_packet) :
-			 udc_rpi_get_ep_buffer_len(ep);
+	bool data_packet_after_setup_is_expected = false;
+	if (state.setup_available) {
+		const struct usb_setup_packet *setup = (const void *)&usb_dpram->setup_packet;
+		if ((setup->wLength != 0U) ||
```

Comment:
Shouldn't the condition here be `&&` instead of `||`?

When `wLength` is 0 then the status stage is always IN ZLP and direction is ignored according to Universal Serial Bus Specification Revision 2.0 9.3.1 bmRequestType
> The state of the Direction bit is ignored if the wLength field is zero, signifying there is no Data stage.

Reply:
Okay, I have refactored the PR a lot, I fixed the issue you mentioned in this thread, and now I'm also covering some conditions that I have not covered before. I have also done more detailed testing on this PR than before.
I have also confirmed that it has a massive positive inpact on issue #52705 (but this issue also needs another change related to data0/data1 sequencing which I will submit as a separate PR very soon).

I have also added the following to the commit message:
    Summary of the changes:
     - Rework the logic around deciding wether to re-arm the out endpoint after a read.
       For non-0 endpoint the previous behaviour is kept, however for EP0 it is only re-
       armed if more OUT transactions are expected for that SETUP transfer (be it data-
       stage or status-stage)
     - Force un-arm the EP0_OUT endpoint in case a stall condition is observed.
     - When a setup transfer is received check if EP0_OUT is already armed. If armed
       then log a warning message, and force-disarm it.